server/src/main/java/org/elasticsearch/rest/action/document/RestBulkAction.java

```diff
@@ -66,38 +76,124 @@ public String getName() {
-        bulkRequest.timeout(request.paramAsTime("timeout", BulkShardRequest.DEFAULT_TIMEOUT));
-        bulkRequest.setRefreshPolicy(request.param("refresh"));
-        bulkRequest.add(
-            request.requiredContent(),
-            defaultIndex,
-            defaultRouting,
-            defaultFetchSourceContext,
-            defaultPipeline,
-            defaultRequireAlias,
-            defaultRequireDataStream,
-            defaultListExecutedPipelines,
-            allowExplicitIndex,
-            request.getXContentType(),
-            request.getRestApiVersion()
-        );
 
-        return channel -> client.bulk(bulkRequest, new RestRefCountedChunkedToXContentListener<>(channel));
+        @Override
+        public void accept(RestChannel restChannel) throws Exception {
+            this.restChannel = restChannel;
+        }
+
+        @Override
+        public void handleChunk(RestChannel channel, ReleasableBytesReference chunk, boolean isLast) {
+            assert channel == restChannel;
+
+            final ReleasableBytesReference data;
+            try {
+                // TODO: Check that the behavior here vs. globalRouting, globalPipeline, globalRequireAlias, globalRequireDatsStream in
+                // BulkRequest#add is fine
+
+                unParsedChunks.add(chunk);
+
+                if (unParsedChunks.size() > 1) {
+                    ReleasableBytesReference[] bytesReferences = unParsedChunks.toArray(new ReleasableBytesReference[0]);
+                    data = new ReleasableBytesReference(
+                        CompositeBytesReference.of(bytesReferences),
+                        () -> Releasables.close(bytesReferences)
+                    );
+                } else {
+                    data = chunk;
+                }
+
+                int bytesConsumed = parser.incrementalParse(
+                    data,
+                    defaultIndex,
+                    defaultRouting,
+                    defaultFetchSourceContext,
+                    defaultPipeline,
+                    defaultRequireAlias,
+                    defaultRequireDataStream,
+                    defaultListExecutedPipelines,
+                    allowExplicitIndex,
+                    request.getXContentType(),
+                    (request, type) -> items.add(request),
+                    items::add,
+                    items::add,
+                    stringDeduplicator
+                );
+
+                accountParsing(bytesConsumed);
+
+            } catch (IOException e) {
+                // TODO: Exception Handling
+                throw new UncheckedIOException(e);
+            }
+
+            if (isLast) {
+                assert unParsedChunks.isEmpty();
+                assert channel != null;
+                handler.lastItems(items, data, new RestRefCountedChunkedToXContentListener<>(channel));
+            } else {
+                handler.addItems(items, data, () -> request.contentStream().next());
```

Comment:
How do you handle case when chunk is smaller than single item? I think "items" might be empty here can be short-circuit to stream.next()